org.lflang/src/org/lflang/generator/c/CUtil.java

```diff
@@ -553,51 +553,68 @@ static public String triggerRefNested(PortInstance port, String runtimeIndex, St
-           } catch (IOException e) {
-               // Files has failed to copy the file, possibly since
-               // it doesn't exist. Will try to find the file as a
-               // resource before giving up.
-           }
-       }
-
-       // Try to copy the file as a resource.
-       // If this is missing, it should have been previously reported as an error.
-       try {
-           String filenameWithoutPath = fileName;
-           int lastSeparator = fileName.lastIndexOf(File.separator);
-           if (lastSeparator > 0) {
-               filenameWithoutPath = fileName.substring(lastSeparator + 1); // FIXME: brittle. What if the file is in a subdirectory?
-           }
-           FileUtil.copyFileFromClassPath(fileName, dstDir.resolve(filenameWithoutPath));
-           return filenameWithoutPath;
-       } catch (IOException ex) {
-           // Ignore. Previously reported as a warning.
-           System.err.println("WARNING: Failed to find file " + fileName);
-       }
-
-       return "";
-   }
+     * Copy the 'fileName' (which also could be a directory name) from the
+     * 'srcDirectory' to the 'destinationDirectory'. This function has a
+     * fallback search mechanism, where if `fileName` is not found in the
+     * `srcDirectory`, it will try to find `fileName` via the following
+     * procedure: 
+     *     1- Search in LF_CLASSPATH. @see findFile() 
+     *     2- Search in CLASSPATH. @see findFile() 
+     *     3- Search for 'fileName' as a resource. That means the `fileName` 
+     *        can be '/path/to/class/resource'. @see java.lang.Class.getResourceAsStream()
+     *
+     * @param fileName Name of the file or directory.
+     * @param srcDir   Where the file or directory is currently located.
+     * @param dstDir   Where the file or directory should be placed.
+     * @return The name of the file or directory in destinationDirectory or an empty string on failure.
+     */
+    public static String copyFileOrResource(String fileName, Path srcDir,
+            Path dstDir) {
+        // Try to copy the file or directory from the file system.
+        Path file = findFileOrDirectory(fileName, srcDir);
+        if (file != null) {
+            Path target = dstDir.resolve(file.getFileName());
+            try {
+                if (Files.isDirectory(file)) {
+                    FileUtil.copyDirectory(file, target);
+                } else if (Files.isRegularFile(file)) {
+                    Files.copy(file, target,
+                            StandardCopyOption.REPLACE_EXISTING);
+                }
+                return file.getFileName().toString();
+            } catch (IOException e) {
+                // Failed to copy the file or directory, most likely
+                // because it doesn't exist. Will try to find it as a
+                // resource before giving up.
+            }
+        }
+
+        String filenameWithoutPath = fileName;
```

Comment:
The following takes the tail of a path (e.g., `foo/bar/baz.lf` yields `baz.lf`; `foo/bar` yields `bar`) and just looks for it on the class path. This doesn't seem right to me. I don't think there are any tests for this, and I think the `FIXME` below alludes to this. I realize this `FIXME` preexists this PR, but I think we should fix and not ignore it.